File: assiist_front_end/lib/core/models/calendar_connection.dart
class CalendarConnection {
  final String provider;
  final String email;
  final String? accessToken;
  final String? idToken;
  final String? createdOn;
  final String? refreshToken;
  final String? tokenExpiry;
  final List<String>? scopes;
  final String? syncStatus;
  final String? syncStatusMessage;
  final String? icsUrl;

  CalendarConnection({
    required this.provider,
    required this.email,
    this.accessToken,
    this.idToken,
    this.createdOn,
    this.refreshToken,
    this.tokenExpiry,
    this.scopes,
    this.syncStatus,
    this.syncStatusMessage,
    this.icsUrl,
  });

  factory CalendarConnection.fromJson(Map<String, dynamic> json) {
    return CalendarConnection(
      provider: json['provider'] as String,
      email: json['email'] as String,
      accessToken: json['access_token'] as String?,
      idToken: json['id_token'] as String?,
      createdOn: json['created_on'] as String?,
      refreshToken: json['refresh_token'] as String?,
      tokenExpiry: json['token_expiry'] as String?,
      scopes: json['scopes'] != null ? List<String>.from(json['scopes']) : null,
      syncStatus: json['sync_status'] as String?,
      syncStatusMessage: json['sync_status_message'] as String?,
      icsUrl: json['ics_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'provider': provider,
      'email': email,
      'access_token': accessToken,
      'id_token': idToken,
      'created_on': createdOn,
      'refresh_token': refreshToken,
      'token_expiry': tokenExpiry,
      'scopes': scopes,
      'sync_status': syncStatus,
      'sync_status_message': syncStatusMessage,
      'ics_url': icsUrl,
    };
  }
}

File: assiist_back_end/models/calendar_connection.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime

class CalendarConnection(BaseModel):
    provider: str
    email: str
    access_token: Optional[str] = None
    id_token: Optional[str] = None # Often specific to Google Sign-In, may not be needed for just API access
    ics_url: Optional[str] = None
    created_on: Optional[str] = None # Should be a datetime string (ISO format)
    
    # New fields for robust token management (especially for Google/OAuth providers)
    refresh_token: Optional[str] = None
    token_expiry: Optional[str] = None # Store as ISO datetime string (e.g., from datetime.isoformat())
    scopes: Optional[List[str]] = None # List of scopes granted

    # Fields to indicate sync health and actions needed
    sync_status: Optional[str] = "active" # e.g., "active", "needs_reauthentication", "error_other", "channel_active", "needs_full_resync"
    sync_status_message: Optional[str] = None

    # --- NEW FIELDS for Google Webhooks ---
    google_channel_id: Optional[str] = None
    google_resource_id: Optional[str] = None
    google_channel_expiration: Optional[datetime] = None # Store expiry time as datetime
    google_sync_token: Optional[str] = None # For incremental sync
    # --- END NEW FIELDS --- 